From 4ecfae0ed8b199d7a040dc02b94ed8baa3d6d15b Mon Sep 17 00:00:00 2001
Message-Id: <4ecfae0ed8b199d7a040dc02b94ed8baa3d6d15b.1423172493.git.feitong.yi@intel.com>
In-Reply-To: <18bb7324f916bffbc75e07ea0ffd3e6e68e7b6d8.1423172493.git.feitong.yi@intel.com>
References: <18bb7324f916bffbc75e07ea0ffd3e6e68e7b6d8.1423172493.git.feitong.yi@intel.com>
From: Gaurav K Singh <gaurav.k.singh@intel.com>
Date: Tue, 13 Jan 2015 20:20:48 +0530
Subject: [PATCH 11/29] MUST_REBASE [VPG]: drm/i915: Pass on display related
 interrupts to ADF driver

In cherryyview_irq_handler if adf driver is enabled call the adf
interrupt handler to handle display related interrupts.

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed

For: GMINL-4443
Change-Id: Ia78f3e3d1d05d19749586081334dbfb4a5678749
Signed-off-by: Gaurav K Singh <gaurav.k.singh@intel.com>
Signed-off-by: Shashank Sharma <shashank.sharma@intel.com>
---
 drivers/gpu/drm/i915/i915_irq.c |   37 +++++++++++++++++++++++++++++++++----
 drivers/gpu/drm/i915/i915_reg.h |    5 +++++
 2 files changed, 38 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_irq.c b/drivers/gpu/drm/i915/i915_irq.c
index 898865f..1839088 100644
--- a/drivers/gpu/drm/i915/i915_irq.c
+++ b/drivers/gpu/drm/i915/i915_irq.c
@@ -2311,7 +2311,8 @@ static irqreturn_t cherryview_irq_handler(int irq, void *arg)
 {
 	struct drm_device *dev = arg;
 	struct drm_i915_private *dev_priv = dev->dev_private;
-	u32 master_ctl, iir;
+	u32 master_ctl, count, iir;
+	u32 mask, pipestat, pipestat_val;
 	irqreturn_t ret = IRQ_NONE;
 
 	for (;;) {
@@ -2325,10 +2326,38 @@ static irqreturn_t cherryview_irq_handler(int irq, void *arg)
 
 		gen8_gt_irq_handler(dev, dev_priv, master_ctl);
 
-		valleyview_pipestat_irq_handler(dev, iir);
+		if (i915.enable_intel_adf) {
+			if (g_adf_ready && i915.disable_display && iir) {
+				unsigned long irqflags;
+				spin_lock_irqsave(&dev_priv->irq_lock,
+						irqflags);
+				intel_adf_context_on_event();
+				spin_unlock_irqrestore(&dev_priv->irq_lock,
+						irqflags);
+			} else {
+				/*
+				 * ADF driver is still initializing, so lets
+				 * ignore dispaly interrupts for now
+				 */
 
-		/* Consume port.  Then clear IIR or we'll miss events */
-		i9xx_hpd_irq_handler(dev);
+				/* FIXME: Display interrupts for Pipe C */
+				count = PIPE_A;
+				while (count < PIPE_C) {
+					mask = PIPESTAT_IIR(count);
+					pipestat = PIPESTAT(count);
+					pipestat_val = I915_READ(pipestat);
+					if (pipestat_val)
+						I915_WRITE(pipestat,
+							pipestat_val | mask);
+					count++;
+				}
+			}
+		} else {
+			valleyview_pipestat_irq_handler(dev, iir);
+
+			/* Consume port. Then clear IIR or we'll miss events */
+			i9xx_hpd_irq_handler(dev);
+		}
 
 		I915_WRITE(VLV_IIR, iir);
 
diff --git a/drivers/gpu/drm/i915/i915_reg.h b/drivers/gpu/drm/i915/i915_reg.h
index 58684df..edad7a2 100644
--- a/drivers/gpu/drm/i915/i915_reg.h
+++ b/drivers/gpu/drm/i915/i915_reg.h
@@ -4044,6 +4044,11 @@ enum punit_power_well {
 #define PIPEFRAME(pipe) _PIPE2(pipe, _PIPEAFRAMEHIGH)
 #define PIPEFRAMEPIXEL(pipe)  _PIPE2(pipe, _PIPEAFRAMEPIXEL)
 #define PIPESTAT(pipe) _PIPE2(pipe, _PIPEASTAT)
+#define PIPESTAT_IIR(pipe) ((pipe == PIPE_A) ? \
+		I915_DISPLAY_PIPE_A_EVENT_INTERRUPT \
+			: (pipe == PIPE_B) \
+				? I915_DISPLAY_PIPE_B_EVENT_INTERRUPT \
+					: I915_DISPLAY_PIPE_C_EVENT_INTERRUPT);
 
 #define _PIPE_MISC_A			0x70030
 #define _PIPE_MISC_B			0x71030
-- 
1.7.9.5

